File: string/string-1/76.py
# Tc = O(m + n)
# Sc = O(n)

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        # Length of the main string
        n = len(s)

        # Build a frequency map (instead of Counter) for all characters in t
        mp = {}
        for ch in t:
            mp[ch] = mp.get(ch, 0) + 1

        # Total number of characters we still need to find
        requiredCount = len(t)

        # Initialize sliding window pointers and minimum window tracking variables
        i = 0
        j = 0
        minStart = 0
        minWindow = float('inf')

        # Expand the right boundary of the window
        while j < n:
            ch_j = s[j]  # Current character at right pointer

            # If current character is still needed, reduce the count of required characters
            if ch_j in mp and mp[ch_j] > 0:
                requiredCount -= 1

            # Decrease frequency count (can go below zero if character appears extra times)
            if ch_j in mp:
                mp[ch_j] -= 1
            else:
                mp[ch_j] = -1  # handle characters not in t

            # When all required characters are matched, try to shrink the window
            while requiredCount == 0:
                # Update the minimum window if the current one is smaller
                if minWindow > j - i + 1:
                    minWindow = j - i + 1
                    minStart = i

                # Try to move the left pointer to shrink the window
                ch_i = s[i]

                # Restore the frequency of the left character since it’s going out of the window
                if ch_i in mp:
                    mp[ch_i] += 1
                    # If this character becomes needed again, increase requiredCount
                    if mp[ch_i] > 0:
                        requiredCount += 1
                i += 1  # move left pointer

            # Move right pointer to expand the window
            j += 1

        # If no valid window found, return empty string
        return "" if minWindow == float('inf') else s[minStart:minStart + minWindow]

File: string/string-1/76.js
class Solution {
  minWindow(s, t) {
    const n = s.length;

    // Step 1: Build a frequency map for all characters in t
    const mp = {};
    for (const ch of t) {
      mp[ch] = (mp[ch] || 0) + 1;
    }

    // Step 2: Initialize variables
    let requiredCount = t.length; // total chars we still need
    let i = 0;                    // left pointer
    let j = 0;                    // right pointer
    let minStart = 0;             // start index of minimum window
    let minWindow = Infinity;     // length of smallest valid window

    // Step 3: Expand the right boundary of the window
    while (j < n) {
      const ch_j = s[j]; // current character at right pointer

      // If character is needed, reduce requiredCount
      if (mp[ch_j] !== undefined && mp[ch_j] > 0) {
        requiredCount--;
      }

      // Decrease the frequency count for the current character
      if (mp[ch_j] !== undefined) {
        mp[ch_j]--;
      } else {
        mp[ch_j] = -1; // for chars not in t
      }

      // Step 4: Try to shrink the window from the left when all chars are matched
      while (requiredCount === 0) {
        // Check if current window is smaller than the previous best
        if (minWindow > j - i + 1) {
          minWindow = j - i + 1;
          minStart = i;
        }

        const ch_i = s[i]; // leftmost character in current window

        // Restore its frequency since it's leaving the window
        if (mp[ch_i] !== undefined) {
          mp[ch_i]++;
          // If character becomes needed again, increment requiredCount
          if (mp[ch_i] > 0) {
            requiredCount++;
          }
        }
        i++; // move left pointer
      }

      // Step 5: Expand right pointer
      j++;
    }

    // Step 6: Return result
    return minWindow === Infinity ? "" : s.substring(minStart, minStart + minWindow);
  }
}

// Example usage:
// const sol = new Solution();
// console.log(sol.minWindow("ADOBECODEBANC", "ABC")); // Output: "BANC"
